docs/design/yarn-v1.md

```diff
@@ -0,0 +1,178 @@
+
+The environment variables specific to the Yarn package manager are prefixed by `YARN_`. When specified, they are [loaded into the registry object](https://github.com/yarnpkg/yarn/blob/7cafa512a777048ce0b666080a24e80aae3d66a9/src/registries/base-registry.js#L125) and then made available to the rest of the Yarn application [via the Config object](https://github.com/yarnpkg/yarn/blob/7cafa512a777048ce0b666080a24e80aae3d66a9/src/config.js#L231). An [example](https://github.com/yarnpkg/yarn/blob/7cafa512a777048ce0b666080a24e80aae3d66a9/src/config.js#L422) of this can be seen for the `YARN_YARN_OFFLINE_MIRROR_PRUNING` (undocumented) environment variable.
+
+Yarn also has support for including [NPM environment variables](https://docs.npmjs.com/cli/v10/using-npm/config#environment-variables) prefixed with `NPM_CONFIG_`.
+
+### Project Configuration
+The [package.json](https://classic.yarnpkg.com/en/docs/package-json) and [yarn.lock](https://classic.yarnpkg.com/en/docs/yarn-lock)
+configuration files must be present in order to process the request.
+
+The node_modules directory should be absent. I'm not certain at the moment how to guard against a user-specified modulesFolder at build-time.
+
+Any additional configuration settings in the repository via [.yarnrc](https://classic.yarnpkg.com/en/docs/yarnrc) or .npmrc
+files should be ignored (at least initially) for the prefetch by specifying [`--no-default-rc`](https://github.com/yarnpkg/yarn/blob/7cafa512a777048ce0b666080a24e80aae3d66a9/src/cli/index.js#L73) for the yarn install command.
+
+We never previously honored user-defined settings with Cachito, but this can be investigated and implemented in Cachi2 in a follow-up. At minimum, the
+limitation should be documented.
+
+### SBOM Generation
+The Yarn 1.x CLI doesn't provide enough information to generate an SBOM, so the yarn.lock and package.json files will need to
+be parsed in order to gather the necessary data. The code used to process these files will need to be
+[imported from or inspired by Cachito](https://github.com/containerbuildsystem/cachito/blob/63b8ec0ea615d114ccfa0d08dc0bec49e60e6a75/cachito/workers/pkg_managers/yarn.py)
+and refactored/improved.
+
+#### Key Points
+ - The yarn.lock file can be parsed with [pyarn](https://github.com/containerbuildsystem/pyarn) like in cachito
+ - Workspaces will need to be [processed separately](https://github.com/containerbuildsystem/cachito/blob/63b8ec0ea615d114ccfa0d08dc0bec49e60e6a75/cachito/workers/pkg_managers/yarn.py#L95) from package.json since they do not appear in the yarn.lock file
+ - Dev dependencies can be [determined similiarly](https://github.com/containerbuildsystem/cachito/blob/63b8ec0ea615d114ccfa0d08dc0bec49e60e6a75/cachito/workers/pkg_managers/yarn.py#L140) to cachito
+ - We will need to ensure that there are no collisions in the names of the archive files being added to the offline mirror
```

Comment:
IIUC, collisions could happen when:
- there are two file dependencies with the exact same name
- the unlikely case where there are two git repos with the same name, and the exact same commit hash exists in both of them

I believe we only need to have an explicit check for the former?

Reply:
Taylor correctly pointed out that a registry dependency could be spoofed if the user declared a file dependency that would result in the same filename when prefetched:

```
{
  "dependencies": {
    "chai": "4.2.0",
    "notchai": "https://some.host.com/archive/chai-4.2.0.tgz",
  }
}
```

So the attack surface is bigger than I expected. Still, we would report all dependencies properly in the SBOM, and the build would very likely fail, so I wonder if it is enough to simply document this limitation for now.

Reply:
The scoping is assuming we'll do a full check to verify collisions, closing this.